sonic-mgmt-common: port to libyang3

bradh352 opened the pull request:
SONiC Libyang3 upgrade Tracking ticket https://github.com/sonic-net/sonic-buildimage/issues/22385 

This PR depends on the other PRs mentioned in the above tracking ticket.

The API changes between libyang1 and libyang3 are drastic.

The cvl test cases are currently passing all tests.

This depends on libyang3's [devel](https://github.com/CESNET/libyang/tree/devel) branch plus this rejected patch:
- https://github.com/CESNET/libyang/pull/2362 -- will never be accepted upstream. libyang maintainer says sonic-yang-mgmt needs to be rearchitected, it is a bad application design to need such a flag.

Or the libyang3 package built via sonic-buildimage which is based on v3.7.8 + all the necessary patches, part of https://github.com/sonic-net/sonic-buildimage/pull/21679

Fixes https://github.com/sonic-net/sonic-buildimage/issues/22385

See comment https://github.com/sonic-net/sonic-mgmt-common/pull/157#issuecomment-4463781541 for the description about supporting both libyang1 and libyang3

bradh352 committed:
port to libyang3

Signed-off-by: Brad House <brad@brad-house.com>

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

bradh352 commented:
@dutta-partha probably going to need your help on this as you are the original author.  I've got the port to libyang3 compiling at least, but I'm not entirely sure how to test it or debug it.  I'm sure there are issues here.

@ganglyu I've seen you make some commits here too, maybe you can help?

I think out of all the libyang3 porting this is the one that is the most complex and that I know the least about.

ganglyu commented:
@bradh352 
This pipeline downloads libyang*.deb from the sonic-buildimage.vs master pipeline artifacts. However, since your changes have not been merged into the master branch, the libyang*3.*.deb file is not available.

    - task: DownloadPipelineArtifact@2
      inputs:
        source: specific
        project: build
        pipeline: 142
        artifact: sonic-buildimage.vs
        runVersion: 'latestFromBranch'
        runBranch: 'refs/heads/$(BUILD_BRANCH)'
        patterns: |
            target/debs/bookworm/libyang*.deb
            target/python-wheels/bookworm/sonic_yang_models*.whl
      displayName: "Download sonic buildimage"

ganglyu commented:
You can follow the pipeline to setup your environment and run unit test.
https://github.com/sonic-net/sonic-mgmt-common/blob/master/azure-pipelines.yml

bradh352 commented:
@ganglyu 

> @bradh352 This pipeline downloads libyang*.deb from the sonic-buildimage.vs master pipeline artifacts. However, since your changes have not been merged into the master branch, the libyang_3._.deb file is not available.

Right, I'm aware of that.  This PR is expected to fail in the SONiC CI system (azure pipelines) until the dependent PRs are merged. 

I was hoping someone more familiar with sonic-mgmt-common could help with testing and validation of this particular PR.  Its by far the most complex port to libyang3 out of all the various sub-projects.

ganglyu commented:
@sachinholla would you please help?

sachinholla commented on azure-pipelines.yml:
```diff
@@ -64,7 +64,7 @@ stages:
 
         # LIBYANG
-        sudo dpkg -i ../target/debs/bookworm/libyang*1.0.73*.deb
+        sudo dpkg -i ../target/debs/bookworm/libyang*3.*.deb
```
Pipeline build for this PR is failing with error:
> dpkg: error: cannot access archive '../target/debs/bookworm/libyang*3.*.deb': No such file or directory

Is the deb path correct? Please fix this so that all CVL tests are exercised and libyang usages are validated.

bradh352 replied:
@sachinholla since https://github.com/sonic-net/sonic-buildimage/pull/21679 is not merged, the debian packages are not uploaded as artifacts that can be retrieved at this point in time.  The azure pipelines therefore cannot be used for validation.  It requires manual compilation and validation in a local environment, which requires that you've installed the 'devel' branch of [libyang](https://github.com/CESNET/libyang/tree/devel).  

I've submitted and have had accepted patches I created that sonic depends on in libyang, but they are not yet in a release version ... this PR does depend on those. (https://github.com/sonic-net/sonic-buildimage/pull/21679 does include said patches already).

ganglyu replied:
@bradh352 sachinholla is the expert of the sonic-mgmt-common repository. You can seek help from him.

sachinholla replied:
Hi Brad, understood, I missed your previous comments.. You can can run the same CVL tests locally using `make && make cvl-test` from your sonic-0mgmt-common repo. Details in my response to your [UMF WG post](https://lists.sonicfoundation.dev/g/umf-working-group/message/87).

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

bradh352 commented:
alright, i've got 2 test vms set up, one to test libyang1 the other to test libyang3 so I can trace the various operations through the test suite.

Tests are definitely failing, however. 

So far I've been able to confirm test initialization through `GetModelListInfo()` (and therefore `getModelChildInfo()`) matches between libyang1 and libyang3 which is good since that appears to account for half (or more) of the libyang calls.

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

bradh352 committed:
update error mappings

Signed-off-by: Brad House <brad@brad-house.com>

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

sachinholla commented:
> looks like most of what is left is error mapping as the current test cases appear to be expecting explicit error codes (some of which have been removed) and error messages (which have drastically changed). I'll need to see if I can use additional metadata to derive the old codes or not, otherwise I'll need to change the test cases themselves.

Please do not change the test cases. Each cvl error code (the one checked in the test case) is important and has certain meaning. They should not be be changed. Can you share the test logs?

sachinholla commented on cvl/cvl_api.go:
```diff
@@ -517,8 +515,6 @@ func (c *CVL) ValidateEditConfig(cfgData []cmn.CVLEditConfigData) (cvlErr CVLErr
 				CVL_LOG(WARNING, "\nValidateEditConfig(): OP_CREATE - Err = %v ", err1)
 			}
 
-			c.yp.SetOperation("CREATE")
```
Did you add any alternative for this??
CVL exposes the current operation through a fixed 'config false' yang node which can be used in the must expressions. This call was initializing that internal node. Many of the test failures could be due to this change. This automatic operation type xpath is an important feature of cvl and cannot be dropped. It is required for supporting incremental changes through REST/gNMI APIs.

bradh352 replied:
Pretty sure this is actually unutilized.  This created a detached libyang lyd_node root separate from anything else.  There was no cross-linking between this node and other roots, it was never imported or merged into any other node.  It seemed like some sort of dangling implementation detail that is no longer in use. 

Do you have an example of where it was used so I can look at those test cases?

sachinholla replied:
There will be many test yangs under sonic-mgmt-common/cvl/testdata/schema which are used by cvl gotests. Many of the test cases would be exercising them. But, may be right about it not being utilized. Cvl does not depend on libyang for must expression validations; so this node may not be needed now.
Easy way to chek is to remove only these 3 SetOperation() calls here and run all the tests with libyang1.

sachinholla commented on cvl/internal/yparser/ly_path.go:
```diff
@@ -105,28 +105,3 @@ type lyPathElem struct {
-// - Leafref "/sonic-port:sonic-port/sonic-port:PORT/sonic-port:ifname" of value "Ethernet668" points to a non-existing leaf
-// - Failed to find "extra" as a sibling to "sonic-acl:aclname"
-var (
-	lyBadValue    = regexp.MustCompile(`[Vv]alue "([^"]*)" `)
```
Is libyang3 providing any better way of finding the target node when there is a validation failure? This was also important for cvl error reporting. There was no such facility in libyang1.x and we had to grep the error message.

bradh352 replied:
Not really.  Get the table and key from the data xpath, and the value from the error message.  Basically same as v1 in that regard ... 

I'm still working on error mapping, but only as they come up in test cases ... right now I'm struggling to understand the leafref issues and why they worked in libyang1 but not libyang3 (put a comment on this ticket if you could look at it)

sachinholla replied:
I think we can safely remove this lyBadValue.

We want to return correct key & field names in the validation error object. In most of the cases libyang data xpath used to contain these case. But in some cases it was not providing the field name -- i guess add node cases. Hence those lyElemPrefix, lyElemSuffix and lyUnknownElem regex are used to extract the field name from the message itself. The lyBadValue is actually optional -- it tries extract the value create a format the message in "Field FF has invalid value VV" format. If you remove lyBadValue, the error message would be "Field FF has invalid value" which is also good enough IMO.

bradh352 replied:
If we remove the regex matching, then I've got to update a bunch of test cases.  I'm basically out of time here, its something that someone else can do later I'm sure if they want to.

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

bradh352 commented:
> > looks like most of what is left is error mapping as the current test cases appear to be expecting explicit error codes (some of which have been removed) and error messages (which have drastically changed). I'll need to see if I can use additional metadata to derive the old codes or not, otherwise I'll need to change the test cases themselves.
> 
> Please do not change the test cases. Each cvl error code (the one checked in the test case) is important and has certain meaning. They should not be be changed. Can you share the test logs?

i've got a lot of the test code mappings sorted out, there's probably still a few more so I left some debug logic in there:
https://github.com/sonic-net/sonic-mgmt-common/pull/157/commits/be4e1dc5aafe431fa3b72de705323dd038db0c9f

I've attached the current log output. There's quite a bit of logging I've added trying to work through the test cases so it might look a little strange.
[cvl-tests.log](https://github.com/user-attachments/files/18913944/cvl-tests.log)

bradh352 commented:
@sachinholla 

After deeper review, the `LYD_OPT_NOEXTDEPS` was previously used to prevent leafref/must/when validations.  This is no longer an option as they removed that flag, and people have asked before and it was shot down to re-add it:
https://github.com/CESNET/libyang/issues/2083

I opened another ticket to see if there is better reception now, since I've been getting my patches accepted upstream that are needed for SONiC already. https://github.com/CESNET/libyang/issues/2355

We are probably going to need to rework this though as I doubt they'll make this change.  I'll be honest, I haven't taken the time to learn what sonic-mgmt-common and its dependents are doing, or why its implementing validators that seem to already be in libyang itself (like `c.validateLeafRef()`).  I'm just trying to migrate everything to libyang3 because of other issues I've hit with libyang1 elsewhere.

I guess the question is, is there any reason not to just retrieve the entire redis config (as json), run it through `lyd_parse_data_mem()` to get a `struct lyd_node *`, then merge in the pending changes (`lyd_merge_siblings()`) before running the validation?

Unless I'm missing something, I'm guessing this decision was made for performance reasons, if that's the case, libyang3 should be considerably faster than libyang1 especially since it precompiles the schema tree and uses better data structures.  So maybe the original concerns are no longer there?

sachinholla commented:
> During schema validation `yparser.ValidateSyntax()`, for `TestValidateEditConfig_Create_Chained_Leafref_DepData_Positive` in the libyang3 port, I am getting an error message

From the error message it appears libyang is trying to evaluate the leafref here.. Cvl implements an in-house leafref, must and when expression evaluator and asks libyang to skip those checks through a flag. I don't recollect the exact name, will check the code and get back to you on Monday

bradh352 commented:
@sachinholla 
> I don't think we can go back and change the cvl design now. My recommendation is to replace libynag with the goyang for data type, syntax and mandatory validations. Not sure if the current cvl maintainer can take it up right now. Can you join the next UMF work group call on next Wednesday/Thursday?? I will try add this topic to the agenda.

I can probably join, I'd just need to know the date/time so I can reserve it.

> A quick fix could be to modify the libyang schema tree -- drop "when" & "must" statements, replace "leafref" with corresponding "type". Are there such APIs in libyang3?

I'm not seeing anything to modify the schema tree in place.  The data structures are all public, and they do provide a function to recompile the schema tree, so its in theory possible ... but probably very brittle from release to release.

bradh352 commented:
@sachinholla  
> Yes, this is the flag I was referring to in my previous response. If libyang3 removed it, then it explains those error messages. CVL author initially planned to use libyang for all the validations but ultimately ended with a golang xpath evaluator for leafref, must & when expression evaluations. Some of the main issues faced with full-libyang approach:
> 
> * Performance and some functional issues
> * Cost of loading the dependent data -- sonic-mgmt-common infra supports incremental yang data updates required for RESTCONF and gNMI. Like updating interface mtu only.
> * Cost of maintaining the libyang's data tree for candidate config (bulk operations or long lived transactions)

I may for now just add a function to pull the entire existing config from redis since its available and then merge in the changset, even if it is really just a single small change to the config. 

It would at least allow these tests to pass and then allow us to determine if the performance hit is acceptable as a stopgap until the code can be migrated to a different solution like goyang.

mssonicbld commented:
/azp run

bradh352 commented:
> @sachinholla
> 
> > Yes, this is the flag I was referring to in my previous response. If libyang3 removed it, then it explains those error messages. CVL author initially planned to use libyang for all the validations but ultimately ended with a golang xpath evaluator for leafref, must & when expression evaluations. Some of the main issues faced with full-libyang approach:
> > 
> > * Performance and some functional issues
> > * Cost of loading the dependent data -- sonic-mgmt-common infra supports incremental yang data updates required for RESTCONF and gNMI. Like updating interface mtu only.
> > * Cost of maintaining the libyang's data tree for candidate config (bulk operations or long lived transactions)
> 
> I may for now just add a function to pull the entire existing config from redis since its available and then merge in the changset, even if it is really just a single small change to the config.
> 
> It would at least allow these tests to pass and then allow us to determine if the performance hit is acceptable as a stopgap until the code can be migrated to a different solution like goyang.

On second thought after looking at the conversion code from CONFIG_DB to the needed yang json, it might be a real pain. There's a whole infrastructure for doing it but it seems deeply tied into some caching stuff and I'm not familiar enough with GOlang to confidently do that (quickly at least, I don't want to spend a lot of time on this).  It might actually be easier to maintain a patch on top of libyang3 to re-introduce a flag similar to `LYD_OPT_NOEXTDEPS` until the libyang usage can be replaced.  

I've already layered on a series of patches to libyang3, two of which have already been accepted upstream, so this is a fairly quick and easy way for me to move this forward ... just realizing this particular patch probably won't ever be upstreamed and we need to actively move forward on removing libyang usages in sonic-mgmt-common.

sachinholla commented:
> It might actually be easier to maintain a patch on top of libyang3 to re-introduce a flag similar to LYD_OPT_NOEXTDEPS until the libyang usage can be replaced.
I've already layered on a series of patches to libyang3, two of which have already been accepted upstream, so this is a fairly quick and easy way for me to move this forward ... just realizing this particular patch probably won't ever be upstreamed and we need to actively move forward on removing libyang usages in sonic-mgmt-common.

I would second this proposal. Sounds much safer than loading the full config_db during every validation call.

Next UMF work group meeting is scheduled for Feb 26th 9PM-10PM PST. You can find the meeting link & details in the [sonic foundation calendar](https://zoom-lfx.platform.linuxfoundation.org/meetings/sonic-foundation) page.

bradh352 commented:
> > It might actually be easier to maintain a patch on top of libyang3 to re-introduce a flag similar to LYD_OPT_NOEXTDEPS until the libyang usage can be replaced.
> 
> I would second this proposal. Sounds much safer than loading the full config_db during every validation call.

I started going down this path, (initial patch here: https://github.com/bradh352/libyang/commit/32a7f74f22d522beb1e5d6581e1ddf96a37f8c3a) 

That does seem to work for preventing errors when a leafref isn't found.

... but speaking with the libyang maintainer, he definitely won't accept anything like this.    I did, however, get him to accept a patch this would depend on with some error propagation: https://github.com/CESNET/libyang/commit/529a594aafd191bb879e13492014136d9be8ddca

Then there's questions about what an "External Dependency" really is at least in the context of our use... is this just cross-module refs we'd fail on, or any non-existent refs even in the same module?  Does this differ for when/must processing?  Or should when/must processing just be completely disabled for libyang?   I'm thinking even in the same module since that part of the tree may not be pulled in by the existing code.

It does really make me want to loop back around on serializing the entire redis config db and feed it in for validation.  I mean redis is in-memory afterall, so it should be fast as long as we use things like SCAN and PIPELINE for the retrieval.  I don't think it makes sense to try to cache it locally, infact I'm not entirely sure why the decision was made to have the `c.tmpDbCache` at all ... I'm not sure of the full purpose as I am still getting familiar with this code, but I don't think this is the pending candidate config cache so it seems like an optimization ... was there any testing done to determine if this was really needed?

It would also mean we could remove all the custom validation logic for leafrefs / must / when ... since that would all be handled by libyang.  I read through the [HLD](https://github.com/sonic-net/SONiC/blob/master/doc/mgmt/Management%20Framework.md) and couldn't find any reference to all this custom validation logic, infact it seems to state that it is intended libyang do this and not the custom Go code.  Of course if the goal is to really replace libyang entirely, I could see this not wanting to be touched, but if we could greatly simplify the code maybe there's no reason to get rid of libyang.

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

bradh352 committed:
getRedisToYangKeys() needs to return an error object

Signed-off-by: Brad House <brad@brad-house.com>

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

bradh352 committed:
fix test cases

Signed-off-by: Brad House <brad@brad-house.com>

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

bradh352 commented:
@sachinholla I've got all test cases passing finally.  I need to upstream some more fixes to libyang and integrate those into sonic-buildimage's patchset until an upstream release is made.

Mind starting to look over the commits?  I've squashed as many as seemed reasonable. 

I left the test case changes as the last commit, they're very minimal and mostly just reflect the error message/code differences where they weren't possible to maintain.  Otherwise I bent over backwards to try to ensure all error codes remained the same.

mssonicbld commented:
/azp run

sachinholla commented on cvl/cvl_syntax.go:
```diff
@@ -127,21 +126,25 @@ func (c *CVL) generateTableFieldsData(config bool, tableName string, jsonNode *j
+				c.appendLeafValue(modelInfo.tableInfo[tableName].mapLeaf[0], jsonFieldNode.Data, &batchInnerListLeafKeys)
+				c.appendLeafValue(modelInfo.tableInfo[tableName].mapLeaf[1], jsonFieldNode.FirstChild.Data, &batchInnerListLeafNonKeys)
+
+				CVL_LOG(WARNING, "1 calling addListNode(%s, %v)", tableName, batchInnerListLeafKeys)
```
why warning??

bradh352 replied:
stale tracing code. removed.

sachinholla commented on cvl/cvl_syntax.go:
```diff
@@ -127,21 +126,25 @@ func (c *CVL) generateTableFieldsData(config bool, tableName string, jsonNode *j
+
+				CVL_LOG(WARNING, "1 calling addListNode(%s, %v)", tableName, batchInnerListLeafKeys)
+				listNode, err := c.addListNode(tableName, parent, tableName, batchInnerListLeafKeys)
+				if listNode == nil {
```
`if err  != nil` would be better

bradh352 replied:
done

sachinholla commented on cvl/cvl_syntax.go:
```diff
@@ -127,21 +126,25 @@ func (c *CVL) generateTableFieldsData(config bool, tableName string, jsonNode *j
+				}
 
-				if errObj := c.yp.AddMultiLeafNodes(modelInfo.tableInfo[tableName].module, listNode, batchInnerListLeaf); errObj.ErrCode != yparser.YP_SUCCESS {
+				CVL_LOG(WARNING, "1 calling AddMultiLeafNodes(%s, %v)", tableName, batchInnerListLeafNonKeys)
```
Is this a temp log added for tracing?? Please cleanup

sachinholla commented on cvl/cvl_syntax.go:
```diff
@@ -209,16 +209,19 @@ func (c *CVL) generateTableData(config bool, jsonNode *jsonquery.Node) (*yparser
 		if !topNodesAdded {
 			// Add top most conatiner e.g. 'container sonic-acl {...}'
-			topNode = c.yp.AddChildNode(modelInfo.tableInfo[tableName].module,
+			topNode, _ = c.yp.AddContainerNode(modelInfo.tableInfo[tableName].module,
```
Why the 2nd output parameter introduced? Can the AddContainerNode() return any error? If yes, better not to ignore it. There is no nil check as well

sachinholla commented on cvl/cvl_test.go:
```diff
@@ -1559,10 +1559,10 @@ func TestValidateEditConfig_Create_Syntax_IncompleteKey_Negative(t *testing.T) {
 
 	verifyValidateEditConfig(t, cfgData, CVLErrorInfo{
-		ErrCode:   cvl.CVL_SYNTAX_MISSING_FIELD,
+		ErrCode:   CVL_SYNTAX_ERROR,
```
Is there any libyang3 limitation for changing specific error to generic error? I was hoping to retain the same errCode, table, key and field names in the error response. Okay to tweak the message a little bit

bradh352 replied:
This change is due to getRedisToYangKeys() failing (the new logic mentioned above).  Previously libyang just took all fields for a list and ran through its validation logic and would emit the first error it hit.  The new version of libyang requires you create a list with the required keys ***first***, then pass in all the remaining list field values after.

If we don't pass all required keys (regardless if the key values are valid) then we get a generic error from libyang that doesn't point to any field at all.

So what I did is checked the key count and if it doesn't match throw an error on the first missing key as the field (or the first excess key as the field).  Once we pass the right number of fields to libyang we can then get a validation failure due to not matching the pattern.

Truthfully I can't see any reason to think the old error was any better at all.  The more useful error is knowing you didn't specify all list keys.

sachinholla commented on cvl/cvl_when_test.go:
```diff
@@ -54,12 +55,11 @@ func TestValidateEditConfig_When_Exp_In_Choice_Negative(t *testing.T) {
 
 	verifyValidateEditConfig(t, cfgDataRule, CVLErrorInfo{
-		ErrCode:   CVL_SEMANTIC_ERROR,
+		ErrCode:   cvl.CVL_SYNTAX_MISSING_FIELD,
```
Please revert this errCode and add the DST_IP to the test payload. Goal of the testcase was to verify the "when" expression which prevents ipv4 configs if IP_TYPE==IPV6. Not sure if there are more test cases to verify such expressions. I think there will be another test case to validate mandatory constraint

bradh352 replied:
ok, wasn't sure. The initial test case was just bad.  I'l make that change and see what happens.

bradh352 replied:
done

sachinholla commented on cvl/internal/yparser/yparser.go:
```diff
@@ -278,9 +373,7 @@ type YParserLeafValue struct {
-	//ctx *YParserCtx    //Parser context
-	root      *YParserNode //Top evel root for validation
-	operation string       //Edit operation
+	dummy string // No members?
```
golang allows empty struct type. You can avoid dummy fields. Doesn't libyang3 need a struct_ly_ctx?

bradh352 replied:
Yeah, a ly_ctx is needed, but the original author made it global rather than a struct/class member for some reason.  I wasn't trying to rearchitect for proper design :)

I'll remove the dummy member.

bradh352 replied:
done

sachinholla commented on cvl/internal/yparser/yparser.go:
```diff
@@ -695,110 +918,140 @@ func getErrorDetails() YParserError {
+//	     ->
+//	  (/sonic-common:operation/sonic-common:operation != 'CREATE') or
+//	  (count(/sonic-interface:sonic-interface/sonic-interface:INTERFACE/sonic-interface:INTERFACE_IPADDR_LIST[sonic-interface:ip_prefix=current()/../ip_prefix] [sonic-interface:portname=(/sonic-vlan:sonic-vlan/sonic-vlan:VLAN_MEMBER/sonic-vlan:VLAN_MEMBER_LIST[sonic-vlan:name=current()]/sonic-vlan:ifname)]) = 0)
+func rewriteXPathPrefix(module *YParserModule, xpath string) string {
```
Can you move to ly_path.go file? This file is already too large. Better to move all path related utilities to ly_path.go

bradh352 replied:
done

sachinholla commented on cvl/internal/yparser/yparser.go:
```diff
@@ -695,110 +918,140 @@ func getErrorDetails() YParserError {
+func rewriteXPathPrefix(module *YParserModule, xpath string) string {
+	hasPrefix := make(map[string]bool)
+
+	re := regexp.MustCompile("[[/][A-Za-z0-9_-]+:")
```
Same regex is compiled for every function call, which can be costly. Make it a global variable

bradh352 replied:
done

sachinholla commented on cvl/internal/yparser/yparser.go:
```diff
@@ -488,56 +602,40 @@ func (yp *YParser) MergeSubtree(root, node *YParserNode) (*YParserNode, YParserE
 	//Just validate syntax
-	if C.lyd_data_validate(&dataTmp, C.LYD_OPT_EDIT|C.LYD_OPT_NOEXTDEPS, (*C.struct_ly_ctx)(ypCtx)) != 0 {
+	if C.lyd_validate_all(&dataPtr, (*C.struct_ly_ctx)(ypCtx), C.LYD_VALIDATE_PRESENT|C.LYD_VALIDATE_NO_STATE|C.LYD_VALIDATE_NOEXTDEPS, nil) != C.LY_SUCCESS {
 		if IsTraceAllowed(TRACE_ONERROR) {
```
Can you point me to the patches to be applied on top of libyang3 master/devel branch to get LYD_VALIDATE_NOEXTDEPS? I will also try these changes on my testbed

bradh352 replied:
yes, that's it.  I also put the link in the description for the ticket.

bradh352 replied:
also, make sure you use the 'devel' branch.  Master needs a few more patches: https://github.com/bradh352/sonic-buildimage/tree/bradh352/libyang3-only/src/libyang3/patch

bradh352 commented on cvl/cvl.go:
```diff
@@ -809,7 +811,16 @@ func getRedisToYangKeys(tableName string, redisKey string) []keyValuePairStruct
 
 	if len(keyNames) != len(keyVals) {
-		return nil //number key names and values does not match
+		cvlErrObj.ErrCode = CVL_SEMANTIC_KEY_INVALID
```
For some background on this, due to error evaluation changes in libyang, we need to detect this ourselves to maintain the level of error detail required, hence needing an error object be returned.

faraazbrcm replied:
Looks good as long as caller is satisfied.

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

sachinholla reviewed (commented):
Hi @bradh352, all tests are passing on my development vm too.. Changes look good to me. Will ask cvl maintainers to take a look.

Seeing a few compiler warnings though:

```
In file included from _cgo_export.c:4:
util.go: In function ‘customLogCb’:
util.go:30:65: warning: passing argument 3 of ‘customLogCallback’ discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
util.go:26:62: note: expected ‘char *’ but argument is of type ‘const char *’
# github.com/Azure/sonic-mgmt-common/cvl/internal/util
../cvl/internal/util/util.go: In function ‘customLogCb’:
../cvl/internal/util/util.go:30:72: warning: passing argument 3 of ‘customLogCallback’ discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
   30 |         customLogCallback(level, (char*)msg, (char*)data_path?data_path:schema_path);
      |                                              ~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~
../cvl/internal/util/util.go:26:62: note: expected ‘char *’ but argument is of type ‘const char *’
   26 | extern void customLogCallback(LY_LOG_LEVEL, char* msg, char* path);
      |                                                        ~~~~~~^~~~
```

bradh352 commented:
> Hi @bradh352, all tests are passing on my development vm too.. Changes look good to me. Will ask cvl maintainers to take a look.
> 
> Seeing a few compiler warnings though:
> 
> ```
> In file included from _cgo_export.c:4:
> util.go: In function ‘customLogCb’:
> util.go:30:65: warning: passing argument 3 of ‘customLogCallback’ discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
> util.go:26:62: note: expected ‘char *’ but argument is of type ‘const char *’
> # github.com/Azure/sonic-mgmt-common/cvl/internal/util
> ../cvl/internal/util/util.go: In function ‘customLogCb’:
> ../cvl/internal/util/util.go:30:72: warning: passing argument 3 of ‘customLogCallback’ discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
>    30 |         customLogCallback(level, (char*)msg, (char*)data_path?data_path:schema_path);
>       |                                              ~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~
> ../cvl/internal/util/util.go:26:62: note: expected ‘char *’ but argument is of type ‘const char *’
>    26 | extern void customLogCallback(LY_LOG_LEVEL, char* msg, char* path);
>       |                                                        ~~~~~~^~~~
> ```

These warnings are from the original code and not something newly introduced.  Its due to the Golang<->C (cgo) stuff, where Go doesn't know about the `const` qualifier in C.  If you really want me to address it I can probably look into it, here's a reference to the issue:
https://stackoverflow.com/questions/32938558/cgo-cant-find-way-to-use-callbacks-with-const-char-argument

bradh352 commented:
@sachinholla any other comments on this review?  If not, could you approve it?

faraazbrcm commented:
@bradh352 Code changes good to me. I wanted to ensure SET/GET going fine using gNMI and RESTCONF interfaces. I will run some some tests next week and update you.

bradh352 commented:
let me know if you need an image to test and for what platform.  I've got a tree with all 17 PRs in this series that I test with here:  https://github.com/bradh352/sonic-buildimage/commits/bradh352/libyang3/

sachinholla commented:
> > ```
> > In file included from _cgo_export.c:4:
> > util.go: In function ‘customLogCb’:
> > util.go:30:65: warning: passing argument 3 of ‘customLogCallback’ discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
> > util.go:26:62: note: expected ‘char *’ but argument is of type ‘const char *’
> > # github.com/Azure/sonic-mgmt-common/cvl/internal/util
> > ../cvl/internal/util/util.go: In function ‘customLogCb’:
> > ../cvl/internal/util/util.go:30:72: warning: passing argument 3 of ‘customLogCallback’ discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
> >    30 |         customLogCallback(level, (char*)msg, (char*)data_path?data_path:schema_path);
> >       |                                              ~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~
> > ../cvl/internal/util/util.go:26:62: note: expected ‘char *’ but argument is of type ‘const char *’
> >    26 | extern void customLogCallback(LY_LOG_LEVEL, char* msg, char* path);
> >       |                                                        ~~~~~~^~~~
> > ```
> 
> These warnings are from the original code and not something newly introduced. Its due to the Golang<->C (cgo) stuff, where Go doesn't know about the `const` qualifier in C. 

@bradh352 -- i don't see this warning with old code. Let us fix it as it shows up during compilation & test runs of all other cvl dependent components. Cgo handles const qualifier if you define a typedef with it. Following works for me:

```diff
diff --git a/cvl/internal/util/util.go b/cvl/internal/util/util.go
index a1664f2..9c357b8 100644
--- a/cvl/internal/util/util.go
+++ b/cvl/internal/util/util.go
@@ -23,7 +23,8 @@ package util
 #cgo LDFLAGS: -lyang
 #include <libyang/libyang.h>

-extern void customLogCallback(LY_LOG_LEVEL, char* msg, char* path);
+typedef const char const_char_t;
+extern void customLogCallback(LY_LOG_LEVEL, const_char_t* msg, const_char_t* path);

 static void customLogCb(LY_LOG_LEVEL level, const char* msg, const char* data_path, const char * schema_path, uint64_t line) {
        (void)line;
@@ -188,7 +189,7 @@ func IsTraceSet() bool {
 changing libyang's global log setting */

 //export customLogCallback
-func customLogCallback(level C.LY_LOG_LEVEL, msg *C.char, path *C.char) {
+func customLogCallback(level C.LY_LOG_LEVEL, msg *C.const_char_t, path *C.const_char_t) {
        if level == C.LY_LLERR {
                CVL_LEVEL_LOG(WARNING, "[libyang Error] %s (path: %s)", C.GoString(msg), C.GoString(path))
        } else {
```

bradh352 commented:
> > > ```
> > > In file included from _cgo_export.c:4:
> > > util.go: In function ‘customLogCb’:
> > > util.go:30:65: warning: passing argument 3 of ‘customLogCallback’ discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
> > > util.go:26:62: note: expected ‘char *’ but argument is of type ‘const char *’
> > > # github.com/Azure/sonic-mgmt-common/cvl/internal/util
> > > ../cvl/internal/util/util.go: In function ‘customLogCb’:
> > > ../cvl/internal/util/util.go:30:72: warning: passing argument 3 of ‘customLogCallback’ discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
> > >    30 |         customLogCallback(level, (char*)msg, (char*)data_path?data_path:schema_path);
> > >       |                                              ~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~
> > > ../cvl/internal/util/util.go:26:62: note: expected ‘char *’ but argument is of type ‘const char *’
> > >    26 | extern void customLogCallback(LY_LOG_LEVEL, char* msg, char* path);
> > >       |                                                        ~~~~~~^~~~
> > > ```
> > 
> > 
> > These warnings are from the original code and not something newly introduced. Its due to the Golang<->C (cgo) stuff, where Go doesn't know about the `const` qualifier in C.
> 
> @bradh352 -- i don't see this warning with old code. Let us fix it as it shows up during compilation & test runs of all other cvl dependent components. Cgo handles const qualifier if you define a typedef with it. Following works for me:
> 
> ```diff
> diff --git a/cvl/internal/util/util.go b/cvl/internal/util/util.go
> index a1664f2..9c357b8 100644
> --- a/cvl/internal/util/util.go
> +++ b/cvl/internal/util/util.go
> @@ -23,7 +23,8 @@ package util
>  #cgo LDFLAGS: -lyang
>  #include <libyang/libyang.h>
> 
> -extern void customLogCallback(LY_LOG_LEVEL, char* msg, char* path);
> +typedef const char const_char_t;
> +extern void customLogCallback(LY_LOG_LEVEL, const_char_t* msg, const_char_t* path);
> 
>  static void customLogCb(LY_LOG_LEVEL level, const char* msg, const char* data_path, const char * schema_path, uint64_t line) {
>         (void)line;
> @@ -188,7 +189,7 @@ func IsTraceSet() bool {
>  changing libyang's global log setting */
> 
>  //export customLogCallback
> -func customLogCallback(level C.LY_LOG_LEVEL, msg *C.char, path *C.char) {
> +func customLogCallback(level C.LY_LOG_LEVEL, msg *C.const_char_t, path *C.const_char_t) {
>         if level == C.LY_LLERR {
>                 CVL_LEVEL_LOG(WARNING, "[libyang Error] %s (path: %s)", C.GoString(msg), C.GoString(path))
>         } else {
> ```

Thanks for letting me know how to resolve that with cgo, i'll go ahead and apply that patch.  Traveling at the moment, but I'll get to it in the next day or two.

bradh352 committed:
cvl/yparser: refactor libyang3 access through yp_*/goly_* helpers

Push every libyang-version-specific call site behind a thin C wrapper in
yparser.go's cgo preamble. The Go code now reaches libyang only through
those wrappers and a small set of accessor helpers, which both shrinks
the surface that has to be aware of libyang's quirks and prepares the
file to be extended with libyang1 support in a follow-up commit -- each
helper becomes the single place where the version difference would land.

This is a pure refactor for the libyang3 build; behavior is unchanged.

Typedefs
--------
Add canonical typedefs over libyang's schema-node, must, when,
ext_instance and per-nodetype subtype structs (yp_snode_t, yp_must_t,
yp_when_t, yp_ext_t, yp_snode_{leaf,leaflist,list,container}_t) so the
rest of the file refers to each type by a single yp_* name. YParserSNode
and getModelChildInfo's parameter switch to *C.yp_snode_t.

Accessor helpers (new)
----------------------
At the end of the cgo preamble, named yp_* to distinguish them from the
older libyang3-flavored golys_* / golysc_* wrappers:

  yp_node_child                    - first child of any inner node
  yp_node_exts_count               - count of extension instances
  yp_node_ext_def_name             - extension def name at index
  yp_node_ext_argument             - extension argument at index
  yp_node_musts                    - must array from leaf/list/etc.
  yp_node_musts_count              - count of musts on a node
  yp_node_must_cond_at             - must xpath expression at index
  yp_node_must_apptag_at           - must apptag at index
  yp_node_must_emsg_at             - must error message at index
  yp_leaf_dflt                     - canonical default value of a leaf
  yp_leaflist_dflts_count          - count of leaf-list defaults
  yp_leaflist_dflt_at              - canonical default at index
  yp_leaflist_min                  - leaf-list min-elements
  yp_list_max                      - list max-elements
  yp_leaf_when_cond                - first when expression on leaf/leaflist
  yp_module_top_container          - top-level container holding data lists
  yp_get_last_error                - last error from a context as one struct

Function wrappers (new)
-----------------------
Lift the remaining libyang calls into wrappers so the Go side does not
spell out libyang's flag enums and out-param conventions directly:

  goly_parse_path                  - lys_parse_path (out-param convention)
  yp_ly_set_loglevel               - ly_log_level
  yp_ly_ctx_destroy                - ly_ctx_destroy
  yp_lyd_free                      - lyd_free_all
  yp_lyd_print_mem                 - lyd_print_mem (LYD_JSON + WITHSIBLINGS)
  yp_lyd_merge                     - lyd_merge_siblings (destruct flag arg)
  yp_lyd_validate_edit             - lyd_validate_all (EDIT/NO_STATE/NOEXTDEPS)

  yp_translate_validation_code     - move the LYVE_* -> YParserRetCode
                                     mapping out of Go and into C, keyed
                                     off a small set of YPC_* mirrors of
                                     the YParser{Ret,Err}Code values

Go migration
------------
getErrorDetails, getModelChildInfo and GetModelListInfo no longer
dereference any libyang-specific subtype fields (sleaf.musts,
sLeafList.dflts, slist.max, mod.compiled.data, ypErrLast.data_path,
etc.); they drive off the new accessors. Initialize, Finish, Debug,
ParseSchemaFile, NodeDump, MergeSubtree, mergeDepData, ValidateSyntax,
FreeNode and translateLYErrToYParserErr drive off the new wrappers.
The .next / .nodetype / .name / .flags direct accesses stay -- those
fields are common across libyang's schema-node structs.

Five previously-used helpers are now redundant and removed:
  golysc_ext_instance_array_count
  golysc_ext_instance_array_idx
  golysc_must_array_count
  golyd_value_array_count
  golyd_value_array_idx

util.go const warning fix
-------------------------
Per sachinholla's review on PR #157, typedef const char as const_char_t
and change customLogCallback's declaration / //export signature to take
const_char_t* -- this lets cgo carry the const qualifier across the Go
boundary so libyang's const char* arguments no longer need to be cast
away, eliminating the -Wdiscarded-qualifiers warning that surfaces in
every build of cvl's cgo glue.

Signed-off-by: Brad House <brad@brad-house.com>

bradh352 committed:
cvl: add libyang1 support alongside libyang3

Wire libyang1 into the yp_*/goly_* abstraction layer introduced in the
previous commit. The yparser package now builds correctly against
either libyang v1 or v3 headers; the choice is purely a question of
which headers are installed, with no Go build tags involved.

The C-side fork is done with
#if defined(LY_VERSION_MAJOR) && LY_VERSION_MAJOR >= 3 guards in the
cgo preamble, and the Go side branches on a single Libyang1 const
derived from a cgo-exposed enum:

  enum { yp_is_libyang1 = 0 };           // libyang3 build
  enum { yp_is_libyang1 = 1 };           // libyang1 build
  const Libyang1 = C.yp_is_libyang1 != 0 // Go-side mirror

yparser.go
----------
* Wrap each typedef in the canonical yp_* set with #if so it points at
  the lysc_* struct under libyang3 and the lys_* struct under libyang1.
* Add libyang1 #else branches for every existing pre-refactor C helper
  (goly_ctx_new, goly_parse_path, golyd_new_inner, golyd_new_list2,
  lyd_multi_new_leaf, lyd_node_leafref_match_in_union,
  golysc_node_list_keys_count, golysc_node_get_when,
  golys_xpath_targets_get) implemented against libyang1's APIs and
  struct layouts. lysc_node_is_union picks up a #if guard to exclude
  it from libyang1 builds (it has no callers there).
* Add libyang1 #else branches for every yp_* accessor: field-name
  renames and array-shape changes (n->exts vs n->ext + n->ext_size,
  lysc_must.cond vs lys_restr.expr, canonical leaf default vs raw
  string on the schema node, when** array vs single when*,
  mod->compiled->data vs mod->data, ly_err_last vs
  ly_err_first->prev) are all hidden inside the accessors.
* Add libyang1 #else branches for the yp_lyd_* / yp_ly_* wrappers.
  The libyang1 validate path explicitly runs
  lyd_check_mandatory_tree before lyd_validate to match libyang3's
  LYD_OPT_EDIT behavior.
* Add a libyang1 #else branch in yp_translate_validation_code for the
  libyang1 LYVE_* enumeration.
* Go-side: AddListNode under libyang1 creates the bare list node and
  then attaches the key leaves via AddMultiLeafNodes, since
  lyd_new_list2 only exists in libyang3.

ly_path.go
----------
* #if-gate golys_module_from_prefix; libyang1 gets a NULL-returning
  stub since rewriteXPathPrefix early-returns under libyang1 (libyang1
  already stores fully-qualified prefixes).
* init() rewrites lyBadValue and lyUnknownElem under libyang1 to
  match its different error-message phrasing ("Failed to find" vs
  "Term node ... not found", etc.).

cvl/internal/util/util.go
-------------------------
* #if-gate the libyang log callback wrapper; libyang1's callback
  signature is 3-arg (level, msg, path) and uses ly_verb / a 2-arg
  ly_set_log_clb, while libyang3 is 5-arg (data_path + schema_path +
  line) and uses ly_log_level / 1-arg ly_set_log_clb.

cvl/cvl_test.go, cvl/cvl_when_test.go
-------------------------------------
* The four tests whose assertions depend on libyang version
  (TestValidateEditConfig_Create_Syntax_IncompleteKey_Negative,
  TestValidateEditConfig_Delete_InvalidKey_Negative,
  TestValidateEditConfig_Update_Semantic_Invalid_Key_Negative and
  TestValidateEditConfig_When_Exp_In_Choice_Negative) now build a
  mutable expected value and override the version-specific fields
  inside `if yparser.Libyang1 { ... }`. The data payload is the
  libyang3-style one in all cases; under libyang1 the validator
  stops at the first invalid field so the assertion just shifts to
  the field libyang1 reports.

Signed-off-by: Brad House <brad@brad-house.com>

bradh352 committed:
TEMP: switch pipeline to install libyang1 for testing

Temporarily revert the azure-pipelines.yml libyang install line back to
the libyang1 .deb so the CI build exercises the libyang1 code path in
yparser/util/ly_path. This commit is intended to be dropped (or its
diff reverted) before merge once we have observed the libyang1 build
pass.

Signed-off-by: Brad House <brad@brad-house.com>

bradh352 committed:
Revert "TEMP: switch pipeline to install libyang1 for testing"

This reverts commit 5a61baf7299c61286e7c81a26b61afc7bee62e8a.

Signed-off-by: Brad House <brad@brad-house.com>

securely1g reviewed (commented):
## Review Notes

Solid engineering work on a genuinely hard migration. A few observations:

**Strengths:**
- The `yp_*` / `goly_*` C abstraction layer is well-designed — hiding all libyang version differences behind ~20 accessor helpers keeps the Go code clean and makes the version fork maintainable from a single location.
- Dual-support via `#if LY_VERSION_MAJOR >= 3` guards + a single Go-side `Libyang1` const is pragmatic. Lets sonic-buildimage switch to libyang3 independently without a big-bang merge.
- The `const_char_t` typedef to fix `-Wdiscarded-qualifiers` is a nice cleanup.

**Concerns / Questions:**
1. **Carried libyang3 patch (NOEXTDEPS):** This is the biggest long-term risk. The patch isn't upstreamable, so every libyang3 version bump requires rebasing it. The exit plan (replace libyang with goyang for syntax validation) makes sense but has no timeline — worth tracking.
2. **Temp CI-toggle commits:** Commits 7-8 appear to be test scaffolding. Should be squashed/removed before final merge to keep history clean.
3. **C helper maintenance surface:** The 20+ new C helpers are now critical path. Consider adding a brief header comment in the cgo preamble documenting the naming convention (`yp_*` = version-abstracted accessors, `goly_*` = function call wrappers) for future contributors.
4. **libyang1 validate path:** The explicit `lyd_check_mandatory_tree` before `lyd_validate` to match libyang3's `LYD_OPT_EDIT` behavior — is there a test that exercises this specifically, or is it covered implicitly by the existing test suite?

**Overall:** Would approve once maintainers confirm integration testing (gNMI/RESTCONF) is clean and the temp commits are cleaned up. The dual-support approach is the right call for the 202605 timeline.
